crates/store/src/accounts/mod.rs

```diff
@@ -0,0 +1,462 @@
+    /// 3. Reconstructing the Merkle path with the historical node values
+    ///
+    /// Returns `None` if the block is in the future or too old (pruned).
+    pub fn open_at(
+        &self,
+        account_id: AccountId,
+        block_number: BlockNumber,
+    ) -> Option<AccountWitness> {
+        match self.historical_state(block_number) {
+            HistoricalState::Latest => Some(self.latest.open(account_id)),
+            HistoricalState::Target(block_number) => {
+                // Ensure overlay exists before reconstruction
+                self.overlays.get(&block_number)?;
+                Self::reconstruct_historical_witness(self, account_id, block_number)
+            },
+            HistoricalState::Future | HistoricalState::TooAncient => None,
+        }
+    }
+
+    /// Gets the account state commitment at the latest block.
+    pub fn get(&self, account_id: AccountId) -> Word {
+        self.latest.get(account_id)
+    }
+
+    /// Checks if the tree contains an account with the given prefix.
+    pub fn contains_account_id_prefix(
+        &self,
+        prefix: miden_objects::account::AccountIdPrefix,
+    ) -> bool {
+        self.latest.contains_account_id_prefix(prefix)
+    }
+
+    /// Determines the historical state of a requested block number.
+    fn historical_state(&self, desired_block_number: BlockNumber) -> HistoricalState {
+        if desired_block_number == self.block_number {
+            return HistoricalState::Latest;
+        }
+
+        // Check if block is in the future
+        if self.block_number.checked_sub(desired_block_number.as_u32()).is_none() {
+            return HistoricalState::Future;
+        }
+
+        // Check if block exists in overlays
+        if !self.overlays.contains_key(&desired_block_number) {
+            return HistoricalState::TooAncient;
+        }
+
+        HistoricalState::Target(desired_block_number)
+    }
+
+    // PRIVATE HELPERS - HISTORICAL RECONSTRUCTION
+    // --------------------------------------------------------------------------------------------
+
+    /// Reconstructs a historical account witness by applying reversion overlays.
+    fn reconstruct_historical_witness(
+        &self,
+        account_id: AccountId,
+        block_target: BlockNumber,
+    ) -> Option<AccountWitness> {
+        // Start with the latest witness
+        let latest_witness = self.latest.open(account_id);
+        let (latest_path, leaf) = latest_witness.into_proof().into_parts();
+        let path_nodes = Self::initialize_path_nodes(&latest_path);
+
+        let leaf_index = NodeIndex::from(leaf.index());
+
+        // Apply reversion overlays to reconstruct historical state
+        let (path, leaf) = Self::apply_reversion_overlays(
+            self.overlays.range(block_target..).rev().map(|(_, overlay)| overlay),
+            path_nodes,
+            leaf_index,
+            leaf,
+        )?;
```

Comment:
nit: I would maybe add an explanation to the comment about why we are reversing the order of overlays.